src/lib/hooks.js

```diff
@@ -0,0 +1,151 @@
+import Triggers from './triggers';
+import CustomFields from './customFields';
+import { store } from '../store';
+import SDK from '../api';
+import { createToken } from '../components/helpers';
+import { loadConfig } from '../lib/main';
+
+const createOrUpdateGuest = async(guest) => {
+	await SDK.grantVisitor({ visitor: { ...guest } });
+	await loadConfig();
+};
+
+const updateIframeGuestData = (data) => {
+	const { iframe, iframe: { guest }, user: _id, token } = store.state;
+	store.setState({ iframe: { ...iframe, guest: { ...guest, ...data } } });
+
+	if (!_id) {
+		return;
+	}
+
+	const guestData = { token, ...data };
+	createOrUpdateGuest(guestData);
+};
+
+const api = {
+	pageVisited(info) {
+		if (info.change === 'url') {
+			Triggers.processRequest(info);
+		}
+
+		const { token, room } = store.state;
+		// TODO: Need to change this behavior... But we need to change the endpoint on backend side before....
+		if (!room) {
+			return;
+		}
+
+		const { change, title, location: { href } } = info;
+		SDK.sendVisitorNavigation({ token, rid: room._id, pageInfo: { change, title, location: { href } } });
+	},
+
+	setCustomField(key, value, overwrite = true) {
+		CustomFields.setCustomField(key, value, overwrite);
+	},
+
+	setTheme(newTheme = {}) {
+		const { iframe, iframe: { theme } } = store.state;
+		if (newTheme.color) {
```

Comment:
create the object first and after call just one `setState` to avoid multiple re renders

Reply:
@ggazzo, fixed!